// @antv/g2plot/lib/interactions/drill-down.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isParentNode = void 0;
var g2_1 = require("@antv/g2");
var util_1 = require("@antv/util");
var drill_down_1 = require("./actions/drill-down");
/**
 * 判断是否为父节点
 */
function isParentNode(context) {
    var data = util_1.get(context, ['event', 'data', 'data'], {});
    return util_1.isArray(data.children) && data.children.length > 0;
}
exports.isParentNode = isParentNode;
g2_1.registerAction('drill-down-action', drill_down_1.DrillDownAction);
g2_1.registerInteraction('drill-down', {
    showEnable: [
        { trigger: 'element:mouseenter', action: 'cursor:pointer', isEnable: isParentNode },
        { trigger: 'element:mouseleave', action: 'cursor:default' },
    ],
    start: [
        {
            trigger: 'element:click',
            isEnable: isParentNode,
            action: ['drill-down-action:click'],
        },
        {
            trigger: 'afterchangesize',
            action: ['drill-down-action:resetPosition'],
        },
    ],
});
