// @antv/g2plot/lib/plots/bar/adaptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptor = void 0;
var tslib_1 = require("tslib");
var adaptor_1 = require("../column/adaptor");
var utils_1 = require("./utils");
var adaptor_2 = require("../column/adaptor");
Object.defineProperty(exports, "meta", { enumerable: true, get: function () { return adaptor_2.meta; } });
/**
 * 柱形图适配器
 * @param params
 */
function adaptor(params) {
    var chart = params.chart, options = params.options;
    var xField = options.xField, yField = options.yField, xAxis = options.xAxis, yAxis = options.yAxis, barStyle = options.barStyle, barWidthRatio = options.barWidthRatio, label = options.label, data = options.data, seriesField = options.seriesField, isStack = options.isStack, minBarWidth = options.minBarWidth, maxBarWidth = options.maxBarWidth;
    // label of bar charts default position is left, if plot has label
    if (label && !label.position) {
        label.position = 'left';
        // 配置默认的 label layout： 如果用户没有指定 layout 和 position， 则自动配置 layout
        if (!label.layout) {
            label.layout = [
                { type: 'interval-adjust-position' },
                { type: 'interval-hide-overlap' },
                { type: 'adjust-color' },
                { type: 'limit-in-plot', cfg: { action: 'hide' } },
            ];
        }
    }
    // 默认 legend 位置
    var legend = options.legend;
    if (seriesField) {
        if (legend !== false) {
            legend = tslib_1.__assign({ position: isStack ? 'top-left' : 'right-top', reversed: isStack ? false : true }, (legend || {}));
        }
    }
    else {
        legend = false;
    }
    // @ts-ignore 直接改值
    params.options.legend = legend;
    // 默认 tooltip 配置
    var tooltip = options.tooltip;
    if (seriesField) {
        if (tooltip !== false) {
            tooltip = tslib_1.__assign({ reversed: isStack ? false : true }, (tooltip || {}));
        }
    }
    // @ts-ignore 直接改值
    params.options.tooltip = tooltip;
    // transpose column to bar
    chart.coordinate().transpose();
    return adaptor_1.adaptor({
        chart: chart,
        options: tslib_1.__assign(tslib_1.__assign({}, options), { label: label, 
            // switch xField and yField
            xField: yField, yField: xField, xAxis: yAxis, yAxis: xAxis, 
            // rename attrs as column
            columnStyle: barStyle, columnWidthRatio: barWidthRatio, minColumnWidth: minBarWidth, maxColumnWidth: maxBarWidth, columnBackground: options.barBackground, 
            // bar 调整数据顺序
            data: utils_1.transformBarData(data) }),
    }, true);
}
exports.adaptor = adaptor;
