// @antv/g2plot/lib/plots/gauge/shapes/meter-gauge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var g2_1 = require("@antv/g2");
// 自定义Shape 部分
g2_1.registerShape('interval', 'meter-gauge', {
    draw: function (cfg, container) {
        // 使用 customInfo 传递参数
        var _a = cfg.customInfo.meter, meter = _a === void 0 ? {} : _a;
        var _b = meter.steps, STEP = _b === void 0 ? 50 : _b, _c = meter.stepRatio, stepRatio = _c === void 0 ? 0.5 : _c;
        var total = this.coordinate.endAngle - this.coordinate.startAngle;
        var interval = total / STEP;
        var gap = 0;
        /**
         * stepRatio 取值范围: (0, 1]
         * 1: interval : gap = stepRatio : (1 - stepRatio)
         * 2: interval * STEP + stepRatio * (STEP - 1) = total
         */
        if (stepRatio > 0 && stepRatio <= 1) {
            interval = total / (((1 - stepRatio) / stepRatio) * (STEP - 1) + STEP);
            gap = (interval * (1 - stepRatio)) / stepRatio;
        }
        var group = container.addGroup();
        // 绘制 gap
        if (gap > 0) {
            var center = this.coordinate.getCenter();
            var radius = this.coordinate.getRadius();
            var _d = g2_1.Util.getAngle(cfg, this.coordinate), startAngle = _d.startAngle, endAngle = _d.endAngle;
            for (var i = startAngle, j = 0; i < endAngle && j < 2 * STEP - 1; j++) {
                var drawn = j % 2;
                if (drawn) {
                    var path = g2_1.Util.getSectorPath(center.x, center.y, radius, i, Math.min(i + gap, endAngle), radius * this.coordinate.innerRadius);
                    group.addShape('path', {
                        name: 'meter-gauge-mask',
                        attrs: {
                            path: path,
                            fill: cfg.color,
                            stroke: cfg.color,
                            lineWidth: 0.5,
                        },
                        // mask 不需要捕捉事件
                        capture: false,
                    });
                }
                i += drawn ? gap : interval;
            }
        }
        return group;
    },
});
